File: python/102/patterns/interpreterpattern.py
class Expression:
    def interpret(self):
        pass

class Number(Expression):
    def __init__(self, value):
        self.value = value

    def interpret(self):
        return self.value

class Add(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self):
        return self.left.interpret() + self.right.interpret()

class Subtract(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self):
        return self.left.interpret() - self.right.interpret()

# Przykład użycia
expr = Add(Number(5), Subtract(Number(10), Number(2)))
print(expr.interpret())  # Output: 13

File: php-oop/patterns/interpreterpattern.php
<?php

interface Expression {
    public function interpret(): int;
}

class Number implements Expression {
    private $value;

    public function __construct(int $value) {
        $this->value = $value;
    }

    public function interpret(): int {
        return $this->value;
    }
}

class Add implements Expression {
    private $left;
    private $right;

    public function __construct(Expression $left, Expression $right) {
        $this->left = $left;
        $this->right = $right;
    }

    public function interpret(): int {
        return $this->left->interpret() + $this->right->interpret();
    }
}

class Subtract implements Expression {
    private $left;
    private $right;

    public function __construct(Expression $left, Expression $right) {
        $this->left = $left;
        $this->right = $right;
    }

    public function interpret(): int {
        return $this->left->interpret() - $this->right->interpret();
    }
}

// Przykład użycia
$expr = new Add(new Number(5), new Subtract(new Number(10), new Number(2)));
echo $expr->interpret(); // Output: 13
?>